File: ios/Classes/Models/HMSRoomAction.swift
//
//  HMSRoom.swift
//  hmssdk_flutter
//
//  Created by govind on 03/02/22.
//

import Foundation
import HMSSDK

class HMSRoomAction {
    static func roomActions(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        switch call.method {
            
        case "get_room":
            getRoom(result, hmsSDK)
            
        case "get_local_peer":
            getLocalPeer(result, hmsSDK)
            
        case "get_remote_peers":
            getRemotePeers(result, hmsSDK)
            
        case "get_peers":
            getPeers(result, hmsSDK)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    static private func getRoom(_ result: FlutterResult, _ hmsSDK: HMSSDK?) {
        
        guard let room = hmsSDK?.room else { result(nil); return }
        
        result(HMSRoomExtension.toDictionary(room))
    }
    
    static private func getLocalPeer(_ result: FlutterResult, _ hmsSDK: HMSSDK?) {
        
        guard let localPeer = hmsSDK?.localPeer else { result(nil); return }
        
        result(HMSPeerExtension.toDictionary(localPeer))
    }
    
    static private func getRemotePeers(_ result: FlutterResult, _ hmsSDK: HMSSDK?) {
        
        guard let peers = hmsSDK?.remotePeers else { result(nil); return }
        
        var listOfPeers = [[String: Any]]()
        peers.forEach { listOfPeers.append(HMSPeerExtension.toDictionary($0)) }
        result(listOfPeers)
    }
    
    static private func getPeers(_ result: FlutterResult, _ hmsSDK: HMSSDK?) {
        
        guard let peers = hmsSDK?.room?.peers else { result(nil); return }
        
        var listOfPeers = [[String: Any]]()
        peers.forEach { listOfPeers.append(HMSPeerExtension.toDictionary($0)) }
        result(listOfPeers)
    }
}

File: android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSRoomAction.kt
package live.hms.hmssdk_flutter
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import live.hms.video.sdk.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class HMSRoomAction {
    companion object {
        fun roomActions(call: MethodCall, result: Result, hmssdk: HMSSDK) {
            when (call.method) {
                "get_room" -> {
                    getRoom(result,hmssdk)
                }
                "get_local_peer" -> {
                    localPeer(result,hmssdk)
                }
                "get_remote_peers" -> {
                    getRemotePeers(result,hmssdk)
                }
                "get_peers" -> {
                    getPeers(result,hmssdk)
                }
                else -> {
                    result.notImplemented()
                }
            }
        }
        private fun getRoom(result: Result,hmssdk: HMSSDK) {
            result.success(HMSRoomExtension.toDictionary(hmssdk?.getRoom()))
        }
        private fun localPeer(result: Result,hmssdk: HMSSDK) {
            result.success(HMSPeerExtension.toDictionary(HMSCommonAction.getLocalPeer(hmssdk)))
        }
        private fun getRemotePeers(result: Result,hmssdk: HMSSDK) {
            val peersList = hmssdk.getRemotePeers()
            val peersMapList = ArrayList<HashMap<String, Any?>?>()
            peersList.forEach {
                peersMapList.add(HMSPeerExtension.toDictionary(it))
            }
            CoroutineScope(Dispatchers.Main).launch {
                result.success(peersMapList)
            }
        }

        // TODO: check behaviour when room is not joined
        private fun getPeers(result: Result,hmssdk: HMSSDK) {
            val peersList = hmssdk.getPeers()
            val peersMapList = ArrayList<HashMap<String, Any?>?>()
            peersList.forEach {
                peersMapList.add(HMSPeerExtension.toDictionary(it))
            }
            CoroutineScope(Dispatchers.Main).launch {
                result.success(peersMapList)
            }
        }
    }
}